{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## In the   Accelerated_Intro_WilkinsonExams    folder\n",
    "\n",
    "## ISSUE THESE COMMADS \n",
    "\n",
    "\n",
    "###   git fetch\n",
    "###   git checkout FETCH_HEAD -- Exam\\ Week\\ 2.ipynb\n",
    "###   git checkout FETCH_HEAD -- Germplasm.tsv\n",
    "###   git checkout FETCH_HEAD -- LocusGene.tsv\n",
    "\n",
    "this will retrieve the exam material without destroying your existing notes\n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "\n",
    "## Start your mySQL Server from a terminal \n",
    "(if it isn't already running)\n",
    "\n",
    "<code>sudo docker start course-mysql</code>\n",
    "<pre>\n",
    "\n",
    "</pre>\n",
    "Don't forget that, if you use sqlMagic, you need to execute the connection lines in your Notebook!\n",
    "\n",
    "<pre>\n",
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "</pre>\n",
    "\n",
    "## Create a new Python3 Jupyter Notebook in your Exam Answers folder\n",
    "\n",
    "commit and push this Notebook when you are finished.\n",
    "\n",
    "You have **2 hours and 50 minutes to complete this exam.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%reload_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## Data Files\n",
    "Germplasm.tsv and LocusGene.tsv contain the datasets we need for the exam.\n",
    "\n",
    "Our objective is to create a database to contain the data in these files, insert the data into the database, then query the database in a variety of ways.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Problem 1:  Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx).  (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "son iguales\n",
      "AT1G01040 AT1G01040\n",
      "son iguales\n",
      "AT1G01060 AT1G01060\n",
      "son iguales\n",
      "AT1G01140 AT1G01140\n",
      "son iguales\n",
      "AT1G01220 AT1G01220\n",
      "son iguales\n",
      "AT2G03720 AT2G03720\n",
      "son iguales\n",
      "AT2G03800 AT2G03800\n",
      "son iguales\n",
      "AT2G04240 AT2G04240\n",
      "son iguales\n",
      "AT2G05210 AT2G05210\n",
      "son iguales\n",
      "AT3G02130 AT3G02130\n",
      "son iguales\n",
      "AT3G02140 AT3G02140\n",
      "son iguales\n",
      "AT3G02230 AT3G02230\n",
      "son iguales\n",
      "AT3G02260 AT3G02260\n",
      "son iguales\n",
      "AT3G02310 AT3G02310\n",
      "son iguales\n",
      "AT3G02680 AT3G02680\n",
      "son iguales\n",
      "AT3G02850 AT3G02850\n",
      "son iguales\n",
      "AT3G02870 AT3G02870\n",
      "son iguales\n",
      "AT3G03260 AT3G03260\n",
      "son iguales\n",
      "AT4G14790 AT4G14790\n",
      "son iguales\n",
      "AT4G15210 AT4G15210\n",
      "son iguales\n",
      "AT4G15560 AT4G15560\n",
      "son iguales\n",
      "AT4G15570 AT4G15570\n",
      "son iguales\n",
      "AT4G15802 AT4G15802\n",
      "son iguales\n",
      "AT4G15880 AT4G15880\n",
      "son iguales\n",
      "AT4G16420 AT4G16420\n",
      "son iguales\n",
      "AT4G16480 AT4G16480\n",
      "son iguales\n",
      "AT5G10480 AT5G10480\n",
      "son iguales\n",
      "AT5G10510 AT5G10510\n",
      "son iguales\n",
      "AT5G11110 AT5G11110\n",
      "son iguales\n",
      "AT5G11260 AT5G11260\n",
      "son iguales\n",
      "AT5G11510 AT5G11510\n",
      "son iguales\n",
      "AT5G12200 AT5G12200\n",
      "son iguales\n",
      "AT5G13290 AT5G13290\n"
     ]
    }
   ],
   "source": [
    "gp = open(\"Germplasm.tsv\", \"r\")\n",
    "lg = open(\"LocusGene.tsv\", \"r\")\n",
    "\n",
    "listA=[]\n",
    "listB=[]\n",
    "for x in gp.readlines():\n",
    "    x = x.rstrip()\n",
    "    #print(\"the current record is\", line)\n",
    "    locus_gp, germplasm, phenotype, pubmed = x.split('\\t')\n",
    "    listA.append(locus_gp)\n",
    "\n",
    "   # print(\"LG:\", locus, \"germplas\", germplasm, \"phenotype:\",phenotype , \"pubmed\",pubmed)\n",
    "for x in lg.readlines():\n",
    "    x = x.rstrip()\n",
    "    #print(\"the current record is\", line)\n",
    "    locus_lg, gene, pl = x.split('\\t')\n",
    "    listB.append(locus_lg)\n",
    "#comparar si son iguales\n",
    "for i in range(1,len(listA)):\n",
    "    if listA[i]==listB[i]:\n",
    "        print(\"son iguales\")\n",
    "        #si se quiere ver los locus se descomenta\n",
    "        print(listA[i] + \" \" + listB[i])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Locus\tgermplasm\tphenotype\tpubmed\n",
      "AT1G01040\tCS3828\tIncreased abundance of miRNA precursors.\t17369351\n",
      "AT1G01060\tlhy-101\tThe mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\t16891401\n",
      "AT1G01140\tSALK_058629\thypersensitive to low potassium media\t17486125\n",
      "AT1G01220\tSALK_012400C\tfkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\t18199744\n",
      "AT2G03720\tSALK_042433\tMultiple straight hairs\t16367956\n",
      "AT2G03800\tgek1-1\tEthanol hypersensitivity.\t15215505\n",
      "AT2G04240\txerico\tResistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\t17933900\n",
      "AT2G05210\tpot1-1\tNo visible phenotype.\t17627276\n",
      "AT3G02130\trpk2-2\tThe homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\t17419837\n",
      "AT3G02140\tafp4-1\tDecreased germination on high concentrations of glucose and sorbitol.\t18484180\n",
      "AT3G02230\trgp1-1\trgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\t21478444\n",
      "AT3G02260\ttir3-1 RGLG1:rglg1 rglg2\tThe triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\t17586653\n",
      "AT3G02310\tsep2-1\tNon-described subtle phenotype.\t10821278\n",
      "AT3G02680\tatnbs1-1\tSignificantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\t17672843\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\n",
      "AT3G02870\tvtc4-1\tascorbate deficient\t16595667\n",
      "AT3G03260\thdg8-1\tNo visible phenotype.\t16778018\n",
      "AT4G14790\tpdd17\tDefective pollen development.\t19237690\n",
      "AT4G15210\tbmy1-2\tPlants cold-shocked for 6h have an increased starch content compared to wildtype.\t16297066\n",
      "AT4G15560\tcla1-1\tMutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\t10982425\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\n",
      "AT4G15802\tAthspb-2\tEarly flowering, reduced fertility, aborted seeds.\t20388662\n",
      "AT4G15880\tesd4-2\tDecreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\t17513499\n",
      "AT4G16420\tprz1-1\tAltered response to auxin and cytokinin\t12747832\n",
      "AT4G16480\tatint4-2\tNo visible phenotype.\t16603666\n",
      "AT5G10480\tpas2-3\tSegregates 25% embryo lethal.\t18799749\n",
      "AT5G10510\tplt3-1\tShort roots and shortened root meristem.\t17960244\n",
      "AT5G11110\tkns2\tDefects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\t18779216\n",
      "AT5G11260\thy5-101\tUnder FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\t16891401\n",
      "AT5G11510\tmyb3r4-1\tNo visible phenotype.\t17287251\n",
      "AT5G12200\tpyd2-2\tThe pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.\t19413687\n",
      "AT5G13290\tcrn-1\tIncreased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.\t12345678\n",
      "\n"
     ]
    }
   ],
   "source": [
    "studentfile = open(\"Germplasm.tsv\", \"r\")\n",
    "print(studentfile.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Locus\tGene\tProteinLength\n",
      "AT1G01040\tDCL1\t332\n",
      "AT1G01060\tLHY\t290\n",
      "AT1G01140\tCIPK9\t223\n",
      "AT1G01220\tFKGP\t190\n",
      "AT2G03720\tMRH6\t189\n",
      "AT2G03800\tGEK1\t196\n",
      "AT2G04240\tXERICO\t256\n",
      "AT2G05210\tPOT1A\t221\n",
      "AT3G02130\tRPK2\t284\n",
      "AT3G02140\tTMAC2\t300\n",
      "AT3G02230\tRGP1\t301\n",
      "AT3G02260\tBIG\t279\n",
      "AT3G02310\tSEP2\t175\n",
      "AT3G02680\tNBS1\t190\n",
      "AT3G02850\tSKOR\t234\n",
      "AT3G02870\tVTC4\t311\n",
      "AT3G03260\tHDG8\t194\n",
      "AT4G14790\tSUV3\t312\n",
      "AT4G15210\tBAM5\t313\n",
      "AT4G15560\tDXS\t219\n",
      "AT4G15570\tMAA3\t294\n",
      "AT4G15802\tHSBP\t254\n",
      "AT4G15880\tESD4\t265\n",
      "AT4G16420\tADA2B\t279\n",
      "AT4G16480\tINT4\t284\n",
      "AT5G10480\tPAS2\t301\n",
      "AT5G10510\tAIL6\t310\n",
      "AT5G11110\tSPS2\t232\n",
      "AT5G11260\tHY5\t221\n",
      "AT5G11510\tMYB3R-4\t336\n",
      "AT5G12200\tPYD2\t310\n",
      "AT5G13290\tCRN\t189\n",
      "\n"
     ]
    }
   ],
   "source": [
    "studentfile = open(\"LocusGene.tsv\", \"r\")\n",
    "print(studentfile.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql DROP DATABASE db1\n",
    "%sql CREATE DATABASE db1\n",
    "%sql USE db1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%sql DROP TABLE Locusgene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql CREATE TABLE Germplasm(locus VARCHAR(20) NOT NULL PRIMARY KEY, germplasm VARCHAR(30) NOT NULL, phenotype VARCHAR(400) NOT NULL, pubmed INTEGER NOT NULL);\n",
    "%sql CREATE TABLE Locusgene(locus VARCHAR(20) NOT NULL PRIMARY KEY, gene VARCHAR(20) NOT NULL,  ProteinLength INTEGER NOT NULL);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT1G01040','CS3828','Increased abundance of miRNA precursors.',17369351)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT1G01060','lhy-101','The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.',16891401)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT1G01140','SALK_058629','hypersensitive to low potassium media',17486125)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT1G01220','SALK_012400C','fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.',18199744)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT2G03720','SALK_042433','Multiple straight hairs',16367956)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT2G03800','gek1-1','Ethanol hypersensitivity.',15215505)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT2G04240','xerico','Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.',17933900)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT2G05210','pot1-1','No visible phenotype.',17627276)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02130','rpk2-2','The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.',17419837)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02140','afp4-1','Decreased germination on high concentrations of glucose and sorbitol.',18484180)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02230','rgp1-1','rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.',21478444)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02260','tir3-1 RGLG1:rglg1 rglg2','The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.',17586653)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02310','sep2-1','Non-described subtle phenotype.',10821278)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02680','atnbs1-1','Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.',17672843)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02850','CS3816','The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.',17568770)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02870','vtc4-1','ascorbate deficient',16595667)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G03260','hdg8-1','No visible phenotype.',16778018)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G14790','pdd17','Defective pollen development.',19237690)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G15210','bmy1-2','Plants cold-shocked for 6h have an increased starch content compared to wildtype.',16297066)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G15560','cla1-1','Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.',10982425)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G15570','maa3','Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.',18772186)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G15802','Athspb-2','Early flowering, reduced fertility, aborted seeds.',20388662)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G15880','esd4-2','Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.',17513499)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G16420','prz1-1','Altered response to auxin and cytokinin',12747832)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G16480','atint4-2','No visible phenotype.',16603666)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G10480','pas2-3','Segregates 25% embryo lethal.',18799749)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G10510','plt3-1','Short roots and shortened root meristem.',17960244)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G11110','kns2','Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.',18779216)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G11260','hy5-101','Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.',16891401)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G11510','myb3r4-1','No visible phenotype.',17287251)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G12200','pyd2-2','The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.',19413687)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G13290','crn-1','Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.',12345678)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='db1',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "        #una funcion que lea los archivos de los documents    \n",
    "        gp.seek(0)     \n",
    "        for line in gp.readlines()[1::]:\n",
    "                line = line.rstrip()\n",
    "                #print(\"the current record is\", line)\n",
    "                var1=line.split('\\t')\n",
    "                #print(var1[0])\n",
    "                sql=\"INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('\"+var1[0]+\"','\"+var1[1]+\"','\"+var1[2]+\"',\"+var1[3]+\")\"\n",
    "                \n",
    "                print(sql)\n",
    "                cursor.execute(sql)\n",
    "                \n",
    "                \n",
    "                \n",
    "                \n",
    "        lg.seek(0)     \n",
    "        for line in lg.readlines()[1::]:\n",
    "                line = line.rstrip()\n",
    "                #print(\"the current record is\", line)\n",
    "                var1=line.split('\\t')\n",
    "                #print(var1[0])\n",
    "                sql=\"INSERT INTO Locusgene(locus, gene, ProteinLength) VALUES ('\"+var1[0]+\"','\"+var1[1]+\"',\"+var1[2]+\")\"\n",
    "                #print(sql)\n",
    "                cursor.execute(sql)\n",
    "           \n",
    "        connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "    connection.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locus</th>\n",
       "        <th>gene</th>\n",
       "        <th>ProteinLength</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'DCL1', 332),\n",
       " ('AT1G01060', 'LHY', 290),\n",
       " ('AT1G01140', 'CIPK9', 223),\n",
       " ('AT1G01220', 'FKGP', 190),\n",
       " ('AT2G03720', 'MRH6', 189),\n",
       " ('AT2G03800', 'GEK1', 196),\n",
       " ('AT2G04240', 'XERICO', 256),\n",
       " ('AT2G05210', 'POT1A', 221),\n",
       " ('AT3G02130', 'RPK2', 284),\n",
       " ('AT3G02140', 'TMAC2', 300),\n",
       " ('AT3G02230', 'RGP1', 301),\n",
       " ('AT3G02260', 'BIG', 279),\n",
       " ('AT3G02310', 'SEP2', 175),\n",
       " ('AT3G02680', 'NBS1', 190),\n",
       " ('AT3G02850', 'SKOR', 234),\n",
       " ('AT3G02870', 'VTC4', 311),\n",
       " ('AT3G03260', 'HDG8', 194),\n",
       " ('AT4G14790', 'SUV3', 312),\n",
       " ('AT4G15210', 'BAM5', 313),\n",
       " ('AT4G15560', 'DXS', 219),\n",
       " ('AT4G15570', 'MAA3', 294),\n",
       " ('AT4G15802', 'HSBP', 254),\n",
       " ('AT4G15880', 'ESD4', 265),\n",
       " ('AT4G16420', 'ADA2B', 279),\n",
       " ('AT4G16480', 'INT4', 284),\n",
       " ('AT5G10480', 'PAS2', 301),\n",
       " ('AT5G10510', 'AIL6', 310),\n",
       " ('AT5G11110', 'SPS2', 232),\n",
       " ('AT5G11260', 'HY5', 221),\n",
       " ('AT5G11510', 'MYB3R-4', 336),\n",
       " ('AT5G12200', 'PYD2', 310),\n",
       " ('AT5G13290', 'CRN', 189)]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT*FROM Locusgene\n",
    "#%sql SELECT*FROM Germplasm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='db1',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "\n",
    "        sql =\"SELECT * FROM Germplasm INNER JOIN Locusgene ON Germplasm.locus = Locusgene.locus\"\n",
    "        cursor.execute(sql)\n",
    "        Germplasm = cursor.fetchall()\n",
    "      #  print(Germplasm)\n",
    "        G=Germplasm[0]\n",
    "        A=list(G.keys())\n",
    "        B=(A[0]+ \"\\t\"+ A[1]+ \"\\t\"+ A[2]+ \"\\t\"+ A[3]+ \"\\t\"+ A[4]+ \"\\t\"+ A[5]+ \"\\t\"+ A[6]+\"\\n\")\n",
    "       # print(B)\n",
    "        REPORT = open(\"report.tsv\", \"w\")\n",
    "        REPORT.write(B)  # the write function, using \\n (newline)\n",
    "        #print(G)\n",
    "        for i in Germplasm:\n",
    "            s=str(i[A[0]] + \"\\t\" + i[A[1]] + \"\\t\" + i[A[2]] + \"\\t\" + str(i[A[3]])+ \"\\t\" + str(i[A[4]])+ \"\\t\" + str(i[A[5]])+ \"\\t\" + str(i[A[6]])+\"\\n\")\n",
    "            #print(s)\n",
    "            REPORT = open(\"report.tsv\", \"a\")\n",
    "            REPORT.write(s)\n",
    "            REPORT.close()\n",
    "      \n",
    "     \n",
    "                    \n",
    "\n",
    "           \n",
    "        connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "    connection.close()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a joined report that only includes the Genes SKOR and MAA3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locus</th>\n",
       "        <th>gene</th>\n",
       "        <th>ProteinLength</th>\n",
       "        <th>locus_1</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT3G02850', 'SKOR', 234, 'AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770)]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM Locusgene INNER JOIN Germplasm ON Germplasm.locus = Locusgene.locus WHERE Locusgene.gene = \"SKOR\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='db1',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "       \n",
    "        #hacer una funcion que tenga como entrada el gen que quiero buscar en la tabla unida\n",
    "        def busq(Gene):\n",
    "            sql =\"SELECT * FROM Locusgene INNER JOIN Germplasm ON Germplasm.locus = Locusgene.locus WHERE Locusgene.gene = '\"+str(Gene)+\"'\"\n",
    "            cursor.execute(sql)\n",
    "            results=cursor.fetchall()\n",
    "            #print(results)\n",
    "            b=\"\"\n",
    "            for i in results:\n",
    "                l=list(i.keys())\n",
    "        \n",
    "                s=len(list(i.keys()))\n",
    "                a=\"\"\n",
    "                for j in range(s):\n",
    "                    a+=str((i[l[j]]))+\"\\t\"\n",
    "            \n",
    "                File1 = open(\"report.tsv\", \"a\")\n",
    "                File1.write(a+\"\\n\")\n",
    "                \n",
    "        HEADERS=\"locus\\tprojects\\tproteinlength\\tlocus-Germplasm\\tphenotype\\tpubmed\\n\"\n",
    "        File1 = open(\"report.tsv\", \"a\")\n",
    "        File1.write(HEADERS)\n",
    "        busq(\"SKOR\")\n",
    "        busq(\"MAA3\")\n",
    "  \n",
    "             \n",
    "        #connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "    connection.close()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>COUNT(*)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>1</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(1,)]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT *FROM Germplasm\n",
    "%sql SELECT COUNT(*) FROM Locusgene INNER JOIN Germplasm ON Germplasm.locus = Locusgene.locus WHERE Locusgene.locus = 'AT1G01040'\n",
    "         \n",
    "#%sql SELECT COUNT(*) AS \"Number Of Matches\" FROM Germplasm RIGHT JOIN Locusgene ON Germplasm.locus = \"AT1G01060\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "#hay que hacer una busqueda de AT1G hasta AT5G, se puede hacer con re\n",
    "#se puede crear una funcion en la que el parametro sea el AT1G\n",
    "#pero solo quiero que me cuente\n",
    "\n",
    "import pymysql.cursors\n",
    "import re\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='db1',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "       \n",
    "        #AQUI HAY QUE HACER UNA FUNCION QUE CREE CORTE LA PRIMERA PARTE DEL GEN Y CREAR UN CICLO\n",
    "        def cort(cromo):\n",
    "            \n",
    "            sql=\"SELECT * FROM Locusgene\"\n",
    "            cursor.execute(sql)\n",
    "            results=cursor.fetchall()\n",
    "            cont=0\n",
    "            suma=0\n",
    "            for i in results:\n",
    "                A=i[\"locus\"]\n",
    "               # L=i[\"ProteinLength\"]\n",
    "                #print(A)\n",
    "                match=re.search((cromo+\"\\w+\"),A)\n",
    "                if match:\n",
    "                    C=(match.group(0))\n",
    "                    cont+=1\n",
    "                   # suma+=L\n",
    "                   # print(suma)\n",
    "                   # print(cont)\n",
    "            return cont\n",
    "        E=\"\"\n",
    "        File1 = open(\"report.tsv\", \"a\")\n",
    "        File1.write(\"CHROMOSOME\\tNUMBER_OF_ENTRIES\\n\")\n",
    "        for i in range(1,6):\n",
    "            cromo=str(\"AT\"+str(i)+\"G\")\n",
    "           # print(cromo)\n",
    "            cort(cromo)\n",
    "            #E=cromo+\"\\t\"+ str(cort(cromo)) \n",
    "            File1 = open(\"report.tsv\", \"a\")\n",
    "            File1.write(cromo+\"\\t\"+ str(cort(cromo))+\"\\n\")\n",
    "           \n",
    "            \n",
    "           # print(E)   \n",
    "            \n",
    "        \n",
    "        connection.commit()\n",
    "          \n",
    "            \n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "    connection.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locus</th>\n",
       "        <th>gene</th>\n",
       "        <th>ProteinLength</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'DCL1', 332),\n",
       " ('AT1G01060', 'LHY', 290),\n",
       " ('AT1G01140', 'CIPK9', 223),\n",
       " ('AT1G01220', 'FKGP', 190),\n",
       " ('AT2G03720', 'MRH6', 189),\n",
       " ('AT2G03800', 'GEK1', 196),\n",
       " ('AT2G04240', 'XERICO', 256),\n",
       " ('AT2G05210', 'POT1A', 221),\n",
       " ('AT3G02130', 'RPK2', 284),\n",
       " ('AT3G02140', 'TMAC2', 300),\n",
       " ('AT3G02230', 'RGP1', 301),\n",
       " ('AT3G02260', 'BIG', 279),\n",
       " ('AT3G02310', 'SEP2', 175),\n",
       " ('AT3G02680', 'NBS1', 190),\n",
       " ('AT3G02850', 'SKOR', 234),\n",
       " ('AT3G02870', 'VTC4', 311),\n",
       " ('AT3G03260', 'HDG8', 194),\n",
       " ('AT4G14790', 'SUV3', 312),\n",
       " ('AT4G15210', 'BAM5', 313),\n",
       " ('AT4G15560', 'DXS', 219),\n",
       " ('AT4G15570', 'MAA3', 294),\n",
       " ('AT4G15802', 'HSBP', 254),\n",
       " ('AT4G15880', 'ESD4', 265),\n",
       " ('AT4G16420', 'ADA2B', 279),\n",
       " ('AT4G16480', 'INT4', 284),\n",
       " ('AT5G10480', 'PAS2', 301),\n",
       " ('AT5G10510', 'AIL6', 310),\n",
       " ('AT5G11110', 'SPS2', 232),\n",
       " ('AT5G11260', 'HY5', 221),\n",
       " ('AT5G11510', 'MYB3R-4', 336),\n",
       " ('AT5G12200', 'PYD2', 310),\n",
       " ('AT5G13290', 'CRN', 189)]"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT*from Locusgene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AT1G\n",
      "AT1G\t258.75\n",
      "AT2G\n",
      "AT2G\t215.5\n",
      "AT3G\n",
      "AT3G\t252.0\n",
      "AT4G\n",
      "AT4G\t277.5\n",
      "AT5G\n",
      "AT5G\t271.2857142857143\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#%sql SELECT AVG(Locusgene.ProteinLength) FROM Locusgene WHERE Locusgene.locus = \n",
    "#hay que hacer una busqueda de AT1G hasta AT5G, se puede hacer con re\n",
    "#se puede crear una funcion en la que el parametro sea el AT1G\n",
    "#pero solo quiero que me cuente\n",
    "\n",
    "import pymysql.cursors\n",
    "import re\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='db1',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "       \n",
    "        #AQUI HAY QUE HACER UNA FUNCION QUE CREE CORTE LA PRIMERA PARTE DEL GEN Y CREAR UN CICLO\n",
    "        def cort(cromo):\n",
    "            \n",
    "            sql=\"SELECT * FROM Locusgene\"\n",
    "            cursor.execute(sql)\n",
    "            results=cursor.fetchall()\n",
    "            cont=0\n",
    "            suma=0\n",
    "            avg=float(0)\n",
    "            for i in results:\n",
    "                A=i[\"locus\"]\n",
    "                L=i[\"ProteinLength\"]\n",
    "                #print(A)\n",
    "                match=re.search((cromo+\"\\w+\"),A)\n",
    "                if match:\n",
    "                    C=(match.group(0))\n",
    "                    cont+=1\n",
    "                    suma+=L\n",
    "                    #print(suma)\n",
    "            avg=(suma/cont)\n",
    "            return avg\n",
    "        E=\"\"\n",
    "        File1 = open(\"report.tsv\", \"a\")\n",
    "        File1.write(\"CHROMOSOME\\tNUMBER_OF_ENTRIES\\n\")\n",
    "        for i in range(1,6):\n",
    "            cromo=str(\"AT\"+str(i)+\"G\")\n",
    "            print(cromo)\n",
    "            R=cort(cromo)\n",
    "            E=cromo+\"\\t\"+ str(cort(cromo)) \n",
    "            File1 = open(\"report.tsv\", \"a\")\n",
    "            File1.write(cromo+\"\\t\"+ str(cort(cromo))+\"\\n\")\n",
    "           \n",
    "            \n",
    "            print(E)   \n",
    "            \n",
    "        \n",
    "        connection.commit()\n",
    "          \n",
    "            \n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "    connection.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "</pre>\n",
    "## Don't forget to commit and push your answers before you leave!\n",
    "\n",
    "It was wonderful to have you in my class!  I hope to see you again soon!\n",
    "\n",
    "Good luck with your careers!!\n",
    "\n",
    "Mark"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
