{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "git fetch\n",
    "git checkout FETCH_HEAD -- Exam\\ Week\\ 2.ipynb\n",
    "git checkout FETCH_HEAD -- Germplasm.tsv\n",
    "git checkout FETCH_HEAD -- LocusGene.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sql extension is already loaded. To reload it, use:\n",
      "  %reload_ext sql\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 1: Controls\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx). (hint: This will help you decide how to load the data into the database)\n",
    "\n",
    "--Answer 1:\n",
    "create a bucle for comparing both of secuences using a conditional\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n",
      "they are the same\n"
     ]
    }
   ],
   "source": [
    "#open the files, just for read\n",
    "gp = open(\"Germplasm.tsv\", \"r\")\n",
    "lg = open(\"LocusGene.tsv\", \"r\")\n",
    "#i want to create a function here which the parameter will be the file\n",
    "def cuts(file1):\n",
    "    seq=[]\n",
    "    for x in file1.readlines()[1:]:\n",
    "        x=x.strip()\n",
    "        A=x.split(\"\\t\")\n",
    "        seq.append(A[0]) \n",
    "        tam=len(seq)\n",
    "    return (seq),tam\n",
    "    \n",
    "\n",
    "gp1,gpL=(cuts(gp))\n",
    "lg1,lgL=cuts(lg)\n",
    "if gpL==lgL: #length is the same\n",
    "    for i in range (gpL):\n",
    "        if gp1[i]==lg1[i]:\n",
    "            print(\"they are the same\")\n",
    "        else:\n",
    "            print(\"it is bad\")\n",
    "else:\n",
    "    print(\"the length is different\")\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 2: Design and create the database.\n",
    "It should have two tables - one for each of the two data files.\n",
    "The two tables should be linked in a 1:1 relationship\n",
    "you may use either sqlMagic or pymysql to build the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql DROP DATABASE db1 #this is for start again :)\n",
    "%sql CREATE DATABASE db1\n",
    "%sql USE db1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create both database\n",
    "%sql CREATE TABLE Germplasm(locus VARCHAR(20) NOT NULL PRIMARY KEY, germplasm VARCHAR(30) NOT NULL, phenotype VARCHAR(400) NOT NULL, pubmed INTEGER NOT NULL);\n",
    "%sql CREATE TABLE Locusgene(locus VARCHAR(20) NOT NULL PRIMARY KEY, gene VARCHAR(20) NOT NULL,  ProteinLength INTEGER NOT NULL);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database. There are a variety of strategies to accomplish this. I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT1G01040','CS3828','Increased abundance of miRNA precursors.',17369351)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT1G01060','lhy-101','The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.',16891401)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT1G01140','SALK_058629','hypersensitive to low potassium media',17486125)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT1G01220','SALK_012400C','fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.',18199744)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT2G03720','SALK_042433','Multiple straight hairs',16367956)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT2G03800','gek1-1','Ethanol hypersensitivity.',15215505)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT2G04240','xerico','Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.',17933900)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT2G05210','pot1-1','No visible phenotype.',17627276)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02130','rpk2-2','The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.',17419837)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02140','afp4-1','Decreased germination on high concentrations of glucose and sorbitol.',18484180)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02230','rgp1-1','rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.',21478444)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02260','tir3-1 RGLG1:rglg1 rglg2','The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.',17586653)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02310','sep2-1','Non-described subtle phenotype.',10821278)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02680','atnbs1-1','Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.',17672843)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02850','CS3816','The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.',17568770)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G02870','vtc4-1','ascorbate deficient',16595667)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT3G03260','hdg8-1','No visible phenotype.',16778018)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G14790','pdd17','Defective pollen development.',19237690)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G15210','bmy1-2','Plants cold-shocked for 6h have an increased starch content compared to wildtype.',16297066)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G15560','cla1-1','Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.',10982425)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G15570','maa3','Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.',18772186)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G15802','Athspb-2','Early flowering, reduced fertility, aborted seeds.',20388662)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G15880','esd4-2','Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.',17513499)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G16420','prz1-1','Altered response to auxin and cytokinin',12747832)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT4G16480','atint4-2','No visible phenotype.',16603666)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G10480','pas2-3','Segregates 25% embryo lethal.',18799749)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G10510','plt3-1','Short roots and shortened root meristem.',17960244)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G11110','kns2','Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.',18779216)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G11260','hy5-101','Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.',16891401)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G11510','myb3r4-1','No visible phenotype.',17287251)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G12200','pyd2-2','The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.',19413687)\n",
      "INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('AT5G13290','crn-1','Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.',12345678)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='db1',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:  \n",
    "        \n",
    "        \n",
    "        gp.seek(0)#BECAUSE I OPENED BEFORE     \n",
    "        for line in gp.readlines()[1::]: #I DONT WANT HEADERS\n",
    "                line = line.rstrip()\n",
    "                var1=line.split('\\t')\n",
    "                sql=\"INSERT INTO Germplasm(locus, germplasm, phenotype, pubmed) VALUES ('\"+var1[0]+\"','\"+var1[1]+\"','\"+var1[2]+\"',\"+var1[3]+\")\"\n",
    "                print(sql)\n",
    "                cursor.execute(sql)\n",
    "                \n",
    "       \n",
    "                \n",
    "                \n",
    "        lg.seek(0)     \n",
    "        for line in lg.readlines()[1::]:\n",
    "                line = line.rstrip()\n",
    "                #print(\"the current record is\", line)\n",
    "                var1=line.split('\\t')\n",
    "                #print(var1[0])\n",
    "                sql=\"INSERT INTO Locusgene(locus, gene, ProteinLength) VALUES ('\"+var1[0]+\"','\"+var1[1]+\"',\"+var1[2]+\")\"\n",
    "                #print(sql)\n",
    "                cursor.execute(sql)\n",
    "           \n",
    "        connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 4: Create reports, written to a file\n",
    "\n",
    "Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MY FILE IS CALL report.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='db1',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "        \n",
    "\n",
    "        sql =\"SELECT * FROM Germplasm INNER JOIN Locusgene ON Germplasm.locus = Locusgene.locus\"\n",
    "        cursor.execute(sql)\n",
    "        Germplasm = cursor.fetchall()\n",
    "      #  print(Germplasm)\n",
    "        G=Germplasm[0]\n",
    "        A=list(G.keys())#i wanna know the keys but keep them in a list, for using them for writting and searching for the info in a dicccionay \n",
    "        var1=\"\"\n",
    "        tam=len(A)\n",
    "        for i in range(1,tam):\n",
    "            var1+=A[i]+\"\\t\"\n",
    "        B=(A[0]+ \"\\t\"+ str(var1)+\"\\n\")\n",
    "        #print(B)\n",
    "        REPORT = open(\"report.tsv\", \"w\")\n",
    "        REPORT.write(B)  # the write function, using \\n (newline)\n",
    "        #print(G)\n",
    "        for i in Germplasm:\n",
    "            var2=\"\"\n",
    "            for j in range(1,tam):\n",
    "                \n",
    "                var2+=str(i[A[j]])+\"\\t\" #this is for not write too much\n",
    "                \n",
    "                \n",
    "            s=str(i[A[0]] + \"\\t\" + str(var2)+\"\\n\")\n",
    "            REPORT = open(\"report.tsv\", \"a\")\n",
    "            REPORT.write(s)\n",
    "            REPORT.close()\n",
    "      \n",
    "     \n",
    "                    \n",
    "\n",
    "           \n",
    "        connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "    connection.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Create a joined report that only includes the Genes SKOR and MAA3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "import pymysql.cursors\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='db1',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "       \n",
    "        # make a function whit a gene like a parameter that i want to search in a inner join table \n",
    "        #hacer una funcion que tenga como entrada el gen que quiero buscar en la tabla unida\n",
    "        HEADERS=\"locus\\tprojects\\tproteinlength\\tlocus-Germplasm\\tphenotype\\tpubmed\\n\"\n",
    "        File1 = open(\"report.tsv\", \"a\")\n",
    "        File1.write(\"\\n\"+HEADERS)\n",
    "        def busq(Gene):\n",
    "            sql =\"SELECT * FROM Locusgene INNER JOIN Germplasm ON Germplasm.locus = Locusgene.locus WHERE Locusgene.gene = '\"+str(Gene)+\"'\"\n",
    "            cursor.execute(sql)\n",
    "            results=cursor.fetchall()\n",
    "            #print(results)\n",
    "            b=\"\"\n",
    "            for i in results:\n",
    "                l=list(i.keys())\n",
    "        \n",
    "                s=len(list(i.keys()))\n",
    "                a=\"\"\n",
    "                for j in range(s):\n",
    "                    a+=str((i[l[j]]))+\"\\t\"\n",
    "            \n",
    "                File1 = open(\"report.tsv\", \"a\")\n",
    "                File1.write(a+\"\\n\")\n",
    "                File1.close()\n",
    "                \n",
    "        \n",
    "        busq(\"SKOR\")\n",
    "        busq(\"MAA3\")\n",
    "   \n",
    "  \n",
    "             \n",
    "        #connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "    connection.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# make a search of bucle(AT1G to AT5G) , maybe with re, using a var\n",
    "#the parameter of the function will be the gene that im searching\n",
    "#the return will be the cont\n",
    "\n",
    "import pymysql.cursors\n",
    "import re\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='db1',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "       \n",
    "        #AQUI HAY QUE HACER UNA FUNCION QUE CREE CORTE LA PRIMERA PARTE DEL GEN Y CREAR UN CICLO\n",
    "        def cort(cromo):\n",
    "            \n",
    "            sql=\"SELECT * FROM Locusgene\"\n",
    "            cursor.execute(sql)\n",
    "            results=cursor.fetchall()\n",
    "            cont=0\n",
    "            suma=0\n",
    "            for i in results:\n",
    "                A=i[\"locus\"]\n",
    "               # L=i[\"ProteinLength\"]\n",
    "                #print(A)\n",
    "                match=re.search((cromo+\"\\w+\"),A)\n",
    "                if match:\n",
    "                    C=(match.group(0))\n",
    "                    cont+=1\n",
    "                   # suma+=L\n",
    "                   # print(suma)\n",
    "                   # print(cont)\n",
    "            return cont\n",
    "        E=\"\"\n",
    "        File1 = open(\"report.tsv\", \"a\")\n",
    "        File1.write(\"CHROMOSOME\\tNUMBER_OF_ENTRIES\\n\")\n",
    "        for i in range(1,6):\n",
    "            cromo=str(\"AT\"+str(i)+\"G\")\n",
    "           # print(cromo)\n",
    "            cort(cromo)\n",
    "            #E=cromo+\"\\t\"+ str(cort(cromo)) \n",
    "            File1 = open(\"report.tsv\", \"a\")\n",
    "            File1.write(cromo+\"\\t\"+ str(cort(cromo))+\"\\n\")\n",
    "            File1.close()\n",
    "            \n",
    "           # print(E)   \n",
    "            \n",
    "        \n",
    "        connection.commit()\n",
    "          \n",
    "            \n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "    connection.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AT1G\n",
      "AT1G\t258.75\n",
      "AT2G\n",
      "AT2G\t215.5\n",
      "AT3G\n",
      "AT3G\t252.0\n",
      "AT4G\n",
      "AT4G\t277.5\n",
      "AT5G\n",
      "AT5G\t271.2857142857143\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#%sql SELECT AVG(Locusgene.ProteinLength) FROM Locusgene WHERE Locusgene.locus = \n",
    "#hay que hacer una busqueda de AT1G hasta AT5G, se puede hacer con re\n",
    "#se puede crear una funcion en la que el parametro sea el AT1G\n",
    "#pero solo quiero que me cuente\n",
    "\n",
    "import pymysql.cursors\n",
    "import re\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='db1',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "       \n",
    "        #AQUI HAY QUE HACER UNA FUNCION QUE CREE CORTE LA PRIMERA PARTE DEL GEN Y CREAR UN CICLO\n",
    "        def cort(cromo):\n",
    "            \n",
    "            sql=\"SELECT * FROM Locusgene\"\n",
    "            cursor.execute(sql)\n",
    "            results=cursor.fetchall()\n",
    "            cont=0\n",
    "            suma=0\n",
    "            avg=float(0)\n",
    "            for i in results:\n",
    "                A=i[\"locus\"]\n",
    "                L=i[\"ProteinLength\"]\n",
    "                #print(A)\n",
    "                match=re.search((cromo+\"\\w+\"),A)\n",
    "                if match:\n",
    "                    C=(match.group(0))\n",
    "                    cont+=1\n",
    "                    suma+=L\n",
    "                    #print(suma)\n",
    "            avg=(suma/cont)\n",
    "            return avg\n",
    "        E=\"\"\n",
    "        File1 = open(\"report.tsv\", \"a\")\n",
    "        File1.write(\"CHROMOSOME\\tAVERAGE\\n\")\n",
    "        for i in range(1,6):\n",
    "            cromo=str(\"AT\"+str(i)+\"G\")\n",
    "            print(cromo)\n",
    "            R=cort(cromo)\n",
    "            E=cromo+\"\\t\"+ str(cort(cromo)) \n",
    "            File1 = open(\"report.tsv\", \"a\")\n",
    "            File1.write(cromo+\"\\t\"+ str(cort(cromo))+\"\\n\")\n",
    "            File1.close()\n",
    "            \n",
    "           \n",
    "            \n",
    "            print(E)   \n",
    "            \n",
    "        \n",
    "        connection.commit()\n",
    "          \n",
    "            \n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "    connection.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
